tests/qos/qos_sai_base.py

```diff
@@ -0,0 +1,828 @@
+            "basicParams": {
+                "router_mac": '' if topo in self.SUPPORTED_T0_TOPOS else dutFacts['ansible_Ethernet0']['macaddress'],
+                "server": duthost.host.options['inventory_manager'].get_host(duthost.hostname).vars['ansible_host'],
+                "port_map_file": ptfPortMapFile,
+                "sonic_asic_type": duthost.facts['asic_type'],
+            }
+        }
+
+    @pytest.fixture(scope='class', autouse=True)
+    def changePtfhostMacAddresses(self, ptfhost):
+        """
+            Change MAC addresses (unique) on PTF host. This class-scope fixture runs once before test start
+ 
+            Args:
+                ptfhost (AnsibleHost): Packet Test Framework (PTF)
+ 
+            Returns:
+                None
+        """
+        ptfhost.script("scripts/change_mac.sh")
+
+    def runPtfTest(self, ptfhost, testCase='', testParams={}):
+        """
+            Runs QoS SAI test case on PTF host
+ 
+            Args:
+                ptfhost (AnsibleHost): Packet Test Framework (PTF)
+                testCase (str): SAI tests test case name
+                testParams (dict): Map of test params required by testCase
+
+            Returns:
+                None
+
+            Raises:
+                RunAnsibleModuleFail if ptf test fails
+        """
+        assert ptfhost.shell(
+            argv = [
+                "ptf",
+                "--test-dir",
+                "saitests",
+                testCase,
+                "--platform-dir",
+                "ptftests",
+                "--platform",
+                "remote",
+                "-t",
+                ";".join(["{}={}".format(k, repr(v)) for k, v in testParams.items()]),
+                "--disable-ipv6",
+                "--disable-vxlan",
+                "--disable-geneve",
+                "--disable-erspan",
+                "--disable-mpls",
+                "--disable-nvgre",
+                "--log-file",
+                "/tmp/{0}.log".format(testCase),
+            ],
+            chdir = "/root",
+        )["rc"] == 0, "Failed when running test '{0}'".format(testCase)
+
+    @pytest.fixture(scope='class')
+    def releaseAllPorts(self, ptfhost, dutTestParams, updateIptables):
+        """
+            Release all paused ports prior to running QoS SAI test cases
+ 
+            Args:
+                ptfhost (AnsibleHost): Packet Test Framework (PTF)
+                dutTestParams (Fixture, dict): DUT host test params
+                updateIptables (Fixture, dict): updateIptables to run prior to releasing paused ports 
+
+            Returns:
+                None
+
+            Raises:
+                RunAnsibleModuleFail if ptf test fails
+        """
+        self.runPtfTest(ptfhost, testCase="sai_qos_tests.ReleaseAllPorts", testParams=dutTestParams["basicParams"])
+
+    @pytest.fixture(scope='class', autouse=True)
+    def populateArpEntries(self, duthost, ptfhost, dutTestParams, dutConfig, releaseAllPorts):
```

Comment:
There is also an arp populate for PTF topology. Didn't see that specified

Reply:
Indeed! i missed that.